### PR TITLE
graph method for PDHMM event groups that unifies finding/merging and overlap/mutual exclusion

### PR DESCRIPTION
**Pull request by davidbenjamin**
@jamesemery The fun begins.  No change in output yet, but a non-trivial change in implementation.

Instead of making preliminary event groups according to overlap, then merging them according to mutually excluded events, this PR does it all in one step while automatically handling transitivity by treating as a matter of finding connected components of a graph whose vertices are events and whose edges are reasons (overlap and mutex) for events to be in the same event group.

All the failures are from WDL tests.  I assume those are not related.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 1]
```diff
@@ -334,6 +305,42 @@ private static List<Event> makeFinalListOfEventsInOrder(final AssemblyResultSet
...
+
+        // edges due to overlapping position
+        for (int e1 = 0; e1 < eventsInOrder.size(); e1++) {
+            final Event event1 = eventsInOrder.get(e1);
+            for (int e2 = e1 + 1; e2 < eventsInOrder.size(); e2++) {
+                final Event event2 = eventsInOrder.get(e2);
+                if (dragenOverlap(event1, event2)) {
+                    graph.addEdge(event1, event2);
+                } else if (event2.getStart() > event1.getEnd() + 1){
+                    break;
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:325]: i mentioned this elsewhere... break statements always give me pause.... consider labeling the inner loop here for maximum clarity

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 2]
```diff
@@ -334,6 +305,42 @@ private static List<Event> makeFinalListOfEventsInOrder(final AssemblyResultSet
...
+        // edges due to mutual exclusion
+        for (final List<Event> excludedGroup : disallowedPairsAndTrios) {
+            graph.addEdge(excludedGroup.get(0), excludedGroup.get(1));
+            if (excludedGroup.size() == 3) {
+                graph.addEdge(excludedGroup.get(1), excludedGroup.get(2));
+            }
+        }
+
+        return new ConnectivityInspector<>(graph).connectedSets().stream()
+                .map(set -> set.stream().sorted(Comparator.comparingInt(Event::getStart)).collect(Collectors.toList()))
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:339]: for consisetncy sake consider using the EventComparator (that is sorting the indels elsewhere in this class) as well...

Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:339]: or possibly let the event group constructor handle the sorting of these events...

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 3]
```diff
@@ -334,6 +305,42 @@ private static List<Event> makeFinalListOfEventsInOrder(final AssemblyResultSet
         return eventsInOrder;
     }
 
+    /**
+     * Partition events into clusters that must be considered together, either because they overlap or because they belong to the
+     * same mutually exclusive pair or trio.  To find this clustering we calculate the connected components of an undirected graph
+     * with an edge connecting events that overlap or are mutually excluded.
+     */
+    private static List<EventGroup> getEventGroupClusters(List<Event> eventsInOrder, List<List<Event>> disallowedPairsAndTrios) {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:313]: I like this... you should keep in mind that you will need to do a similar grouping operation based off of STR adjacency in the future as well... (looks easy enough to add here...)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java [hunk 4]
```diff
@@ -903,7 +910,12 @@ private static double dragenStart(final Event event) {
         return event.getStart() + (event.isIndel() ? (event.isSimpleDeletion() ? 1 : 0.5) : 0);
     }
 
-    private static void eventGroupsMessage(final Haplotype referenceHaplotype, final boolean debug, final Map<Double, List<Event>> eventsByDRAGENCoordinates) {
+    private static final boolean dragenOverlap(final Event event1, final Event event2) {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:913]: javadoc this to be explicit about what and why the boundaries are here (for example we allow compound events here) also rename it

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 2 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:339]: using the Comparator now.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 4 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:913]: moot due to previous PR

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java (hunk 1 above)
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/PartiallyDeterminedHaplotypeComputationEngine.java:325]: I eliminated the `break` by putting the condition into the `for` loop statement.

**Comment by davidbenjamin** [comment]
@jamesemery Ready for re-review.

 I ran this branch against master in `--dragen-378-concordance-mode` on a PCR plus NA12878 sample over all of chromosome 20.  The output was completely identical and runtime appeared unaffected.  If anything I think this branch had a tendency to be faster on all of the longest-running of the 100-way scattered jobs.

**Review by jamesemery — APPROVED** [review]
Thank you for double checking the outputs on a larger section. 👍

**Pull request merged**